sourcetool/pkg/sourcetool/implementation.go

```diff
@@ -54,3 +102,325 @@ func (impl *defaultToolImplementation) GetActiveControls(opts *Options) (slsa.Co
+// to the repository
+func (impl *defaultToolImplementation) CreateWorkflowPR(opts *Options) error {
+	// Branchname to be created on the user's fork
+	branchname := fmt.Sprintf("slsa-source-workflow-%d", time.Now().Unix())
+
+	// Check Environment
+	gh := kgithub.New()
+
+	userForkOrg := opts.UserForkOrg
+	userForkRepo := opts.Repo // For now we only support forks with the same name
+
+	// Clone the repository being onboarded
+	gitCloneOpts := &gogit.CloneOptions{Depth: 1}
+	repo, err := kgithub.PrepareFork(
+		branchname, opts.Owner, opts.Repo,
+		userForkOrg, userForkRepo,
+		opts.UseSSH, opts.UpdateRepo, gitCloneOpts,
+	)
+	if err != nil {
+		return fmt.Errorf("while preparing the repository fork: %w", err)
+	}
+
+	defer func() {
+		repo.Cleanup() //nolint:errcheck,gosec
+	}()
+
+	// Create the workflow file here
+	fullPath := filepath.Join(repo.Dir(), workflowPath)
+	if err := os.MkdirAll(filepath.Dir(fullPath), os.FileMode(0o755)); err != nil {
+		return fmt.Errorf("creating workflow directory: %w", err)
+	}
+
+	// Write the workflow file to disk
+	if err := os.WriteFile(fullPath, []byte(workflowData), os.FileMode(0o644)); err != nil {
+		return fmt.Errorf("writing workflow data to disk: %w", err)
+	}
+
+	// add the modified manifest to staging
+	logrus.Debugf("Adding %s to staging area", workflowPath)
+	if err := repo.Add(workflowPath); err != nil {
+		return fmt.Errorf("adding workflow file to staging area: %w", err)
+	}
+
+	commitMessage := "Add SLSA Source attesting workflow"
+
+	// Create the commit
+	if err := repo.UserCommit(commitMessage); err != nil {
+		return fmt.Errorf("committing changes to workflow: %w", err)
+	}
+
+	// Push commit to branch in the user's fork
+	logrus.Infof("Pushing workflow commit to %s/%s", userForkOrg, userForkRepo)
+	if err := repo.PushToRemote(kgithub.UserForkName, branchname); err != nil {
+		return fmt.Errorf("pushing %s to %s/%s: %w", kgithub.UserForkName, userForkOrg, userForkRepo, err)
+	}
+
+	prBody := `This pull request adds a workflow to the repository to attest the
+SLSA Source compliance on every push.
+`
+	// Create the Pull Request
+	pr, err := gh.CreatePullRequest(
+		opts.Owner, opts.Repo, opts.Branch,
+		fmt.Sprintf("%s:%s", userForkOrg, branchname),
+		commitMessage, prBody, false,
+	)
+	if err != nil {
+		return fmt.Errorf("creating the pull request in %s: %w", opts.Owner, err)
+	}
+	logrus.Infof(
+		"Successfully created PR: %s%s/%s/pull/%d",
+		kgithub.GitHubURL, opts.Owner, opts.Repo, pr.GetNumber(),
+	)
+
+	// Success!
+	return nil
+}
+
+func (impl *defaultToolImplementation) CheckPolicyFork(opts *Options) error {
+	policyOrg, policyRepo, ok := strings.Cut(opts.PolicyRepo, "/")
+	if !ok || policyRepo == "" {
+		return fmt.Errorf("unable to parse policy repository slug")
+	}
+	userForkOrg := opts.UserForkOrg
+	userForkRepo := policyRepo // For now we only support forks with the same name
+
+	// Check the user has a fork of the slsa repo
+	if err := kgithub.VerifyFork(
```

Comment:
Maybe it's too much now, but could we create forks for people if they don't already have them?  Getting them to create the right forks might be tricky.